tst/testinstall/MatrixObj/testmatobj_new.g

```diff
@@ -0,0 +1,422 @@
+#    opt.customGenerator is bound and opt.invertible is bound:
+#       opt.invertible is used to generate matrices and therefore must be bound 
+#       to a function accordingly
+
+MatObjTest_GenerateExample := function(filter, opt_in)
+    local inv, dims, dim, doms, dom, generator, examples, ex, getSourceOfTruth, result, _case, cases, opt;
+    
+    examples := [];
+    opt := ShallowCopy(opt_in);
+
+    if not IsBound(opt.cases) then 
+        if IsBound(opt.dimensions) then 
+            dims := opt.dimensions;
+        else 
+            dims := [[1,1],[2,2],[3,3],[2,5],[5,5],[1,3],[4,1],[6,2]];
+        fi;
+
+        if IsBound(opt.domains) then 
+            doms := opt.domains;
+        else
+            doms := [Integers, Rationals, GF(2), GF(5), GF(49)];
+        fi;
+
+        if IsBound(opt.invertible) then 
+            doms := Filtered(doms, IsField);
+        fi;
+
+        cases := [];
+
+        for dim in dims do
+            for dom in doms do 
+                Add(cases, [dim[1],dim[2],dom]);
+            od;
+        od;
+    else 
+        cases := opt.cases;
+    fi;
+
+    # set generator to use provided customGenerators if applicable
+    if IsBound(opt.customGenerator) then 
+        if IsBound(opt.invertible) then 
+            # use a small wrapper to call the generator, note that this will
+            # generate invertible nxn matrices where n is the number of rows,
+            # i.e. the number of columns is simply ignored 
+            generator := function(filter, randomSource, nrRows, nrCols, dom) return opt.invertible(filter, randomSource, nrRows, dom); end;
+        else 
+            # here a wrapper is unnecessary since the customGenerator must have
+            # the signature we use below
+            generator := opt.customGenerator;
+        fi;
+    else 
+        if IsBound(opt.invertible) then 
+            generator := function(filter, randomSource, nrRows, nrCols, dom) return RandomInvertibleMat(randomSource, nrRows, dom); end;
+        else 
+            generator := function(filter, randomSource, nrRows, nrCols, dom) return RandomMat(randomSource, nrRows, nrCols, dom); end;
+        fi;
+    fi;
+    
+    for _case in cases do 
+        Add(examples, [generator(filter, opt.randomSource, _case[1], _case[2], _case[3]), _case[1], _case[2], _case[3]]);
+    od;
+
+    if IsBound(opt.sourceOfTruthFilter) then 
+        getSourceOfTruth := x -> NewMatrix(opt.sourceOfTruthFilter, x[4], x[3], x[1]);
+    else 
+        getSourceOfTruth := x -> x[1];
+    fi;
+ 
+    result := [];
+
+    for ex in examples do
+        Add(result, rec(matObj := NewMatrix(filter, ex[4], ex[3], ex[1]),
+            mat := ex[1],
+            sourceOfTruth := getSourceOfTruth(ex)));
+        if BaseDomain(Last(result).matObj) <> ex[4] or NumberRows(Last(result).matObj) <> ex[2] or NumberColumns(Last(result).matObj) <> ex[3] or ConstructingFilter(Last(result).matObj) <> filter then 
```

Comment:
IMHO it'd be more readable if your store the generated matrix in its own local variable (which would have to be added to the `local` statement at the top, of course). E.g. like this
```suggestion
        matObj := NewMatrix(filter, ex[4], ex[3], ex[1]);
        Add(result, rec(matObj := matObj,
            mat := ex[1],
            sourceOfTruth := getSourceOfTruth(ex)));
        if BaseDomain(matObj) <> ex[4] or NumberRows(matObj) <> ex[2] or NumberColumns(matObj) <> ex[3] or ConstructingFilter(matObj) <> filter then 
```